### PR TITLE
Component/CloudUploader: support uploading an array of blobs

### PR DESCRIPTION
**Pull request by hzoo**
https://github.com/behance/issues/issues/3567

Sending the lightroom image blobs through cloudthrough

Poet code will be something like

```js
export function uploadBlobs(config, blobs, options) {
  return CloudUploader.promiseBlobs(getOptions(config, options), blobs);
}

export function upload(config, options) {
  return CloudUploader.promise(getOptions(config, options));
}
```

### CODE REVIEW

File: test/specs/Component/CloudUploader.js [hunk 1]
```diff
@@ -186,6 +186,16 @@ define([
     });
 
     describe('#promise', function() {
+      it('returns a rejected promise with an incorrect config', function(done) {
+        // use the imported/not overwritten CloudUploader
+        spyOn(CloudUploader.prototype, 'addFiles');
```
Comment (mikesherov) [inline test/specs/Component/CloudUploader.js]: does this need to be spied?

File: Component/CloudUploader.js [hunk 2]
```diff
@@ -426,11 +427,38 @@ define([
...
+      } catch (e) {
+        return Promise.reject(e);
+      }
+
+      var promise = uploader.promisify();
+      uploader.choose();
+
+      return promise;
+    },
+    promiseBlobs: function(options, blobs) {
```
Comment (mikesherov) [inline Component/CloudUploader.js]: can we call this `promiseFromFiles`?

File: test/specs/Component/CloudUploader.js (hunk 1 above)
Comment (hzoo) [inline test/specs/Component/CloudUploader.js]: No I don't think so will try removing

File: test/specs/Component/CloudUploader.js [hunk 3]
```diff
@@ -185,14 +185,83 @@ define([
...
+          done();
+        });
+
+        CloudUploader.promiseFromFiles({}).catch(function(e) {
+          expect(e).toEqual(new Error('Please provide a proper `request` configuration property'));
+          done();
+        });
+      });
+
+      iit('returns a promise resolved with an array of progress aware subpromises for blob uploads', function(done) {
```
Comment (mikesherov) [inline test/specs/Component/CloudUploader.js]: It

**Comment by mikesherov** [comment]
Move the tests into a `#promiseFromFiles` describe so as not to break convention.

**Comment by mikesherov** [comment]
:+1:

File: Component/CloudUploader.js [hunk 4]
```diff
@@ -426,11 +427,38 @@ define([
...
+      } catch (e) {
+        return Promise.reject(e);
+      }
+
+      var promise = uploader.promisify();
+      uploader.choose();
+
+      return promise;
+    },
+    promiseFromFiles: function(options, blobs) {
```
Comment (mrjoelkemp) [inline Component/CloudUploader.js]: Why not just pass the blobs as a second arg to `promise` then have the existence of the blobs determine if we should trigger choose or addFiles?

Comment (mrjoelkemp) [inline Component/CloudUploader.js]: The code for both methods is identical except for the action to take

File: test/specs/Component/CloudUploader.js [hunk 5]
```diff
@@ -185,7 +185,25 @@ define([
       });
     });
 
+    describe('#addFiles', function() {
+      it('calls fineUploader\'s addFiles method', function() {
```
Comment (mrjoelkemp) [inline test/specs/Component/CloudUploader.js]: Test for a delegated method isn't valuable

Comment (mrjoelkemp) [inline test/specs/Component/CloudUploader.js]: It's all wiring

File: Component/CloudUploader.js (hunk 4 above)
Comment (mikesherov) [inline Component/CloudUploader.js]: Great idea @mrjoelkemp

File: test/specs/Component/CloudUploader.js [hunk 6]
```diff
@@ -185,7 +185,25 @@ define([
...
+
+        var files = [];
+        this.uploader.addFiles(files);
+        expect(CloudUploader.prototype.addFiles).toHaveBeenCalledWith(files);
+      });
+    });
+
     describe('#promise', function() {
+      it('returns a rejected promise with an incorrect config', function(done) {
+        // use the imported/not overwritten CloudUploader
```
Comment (mrjoelkemp) [inline test/specs/Component/CloudUploader.js]: It's more helpful to say why. The test-extended version stubs out `init`, which this test needs.

File: test/specs/Component/CloudUploader.js [hunk 7]
```diff
@@ -185,7 +185,25 @@ define([
...
+      it('calls fineUploader\'s addFiles method', function() {
+        spyOn(CloudUploader.prototype, 'addFiles');
+
+        var files = [];
+        this.uploader.addFiles(files);
+        expect(CloudUploader.prototype.addFiles).toHaveBeenCalledWith(files);
+      });
+    });
+
     describe('#promise', function() {
```
Comment (mrjoelkemp) [inline test/specs/Component/CloudUploader.js]: This should be `.promise` no?

File: test/specs/Component/CloudUploader.js [hunk 8]
```diff
@@ -225,6 +243,62 @@ define([
           });
         });
       });
+    })
+
+    describe('#promiseFromFiles', function() {
```
Comment (mrjoelkemp) [inline test/specs/Component/CloudUploader.js]: Should be `.promiseFromFiles` no?

File: test/specs/Component/CloudUploader.js (hunk 5 above)
Comment (hzoo) [inline test/specs/Component/CloudUploader.js]: yeah it doesn't do anything

File: test/specs/Component/CloudUploader.js [hunk 9]
```diff
@@ -225,6 +243,62 @@ define([
...
+
+        Uploader.promiseFromFiles({}, blobs).then(function(fileArray) {
+          expect(fileArray[0].file.readerData).toBeDefined();
+          expect(fileArray[1].file).toBe(null);
+          Promise.all([
+            fileArray[0].promise,
+            fileArray[1].promise.catch(function() {
+              return Promise.resolve('caught');
+            }),
+            new Promise(function(resolve) {
```
Comment (mrjoelkemp) [inline test/specs/Component/CloudUploader.js]: A comment above explaining why you need to do this would be helpful

Comment (hzoo) [inline test/specs/Component/CloudUploader.js]: I used the other [test](https://github.com/behance/BeFF/pull/113/files#diff-a7ecd36ee6b1dabc7230b1f391f399acR235) - Promise.all takes a list of promises so I guess it's needed to resolve progress correctly?

Comment (mikesherov) [inline test/specs/Component/CloudUploader.js]: So that the we can easily detect that progress has fired at least once.

**Comment by hzoo** [comment]
Anything else for this?

**Comment by mikesherov** [comment]
Yes, revert the promisify stuff we did. The only change needed was promise taking an additional arg. :-/

File: Component/CloudUploader.js [hunk 10]
```diff
@@ -382,7 +386,7 @@ define([
       });
     }
   }, {
-    promise: function(options) {
+    promise: function(options, blobs) {
```
Comment (mikesherov) [inline Component/CloudUploader.js]: Call this param `files`

**Comment by mikesherov** [comment]
:+1:

**Comment by mrjoelkemp** [comment]
:+1:

**Pull request merged**